contrib/runners/winrm_runner/winrm_runner/winrm_base.py

```diff
@@ -0,0 +1,301 @@
+        status_code = response.status_code
+        if response.timeout:
+            status = action_constants.LIVEACTION_STATUS_TIMED_OUT
+            status_code = WINRM_TIMEOUT_EXIT_CODE
+        elif not succeeded:
+            status = action_constants.LIVEACTION_STATUS_FAILED
+
+        # create result
+        result = {
+            'failed': not succeeded,
+            'succeeded': succeeded,
+            'return_code': status_code,
+            'stdout': response.std_out,
+            'stderr': response.std_err
+        }
+
+        # automatically convert result stdout/stderr from JSON strings to
+        # objects so they can be used natively
+        return (status, jsonify.json_loads(result, RESULT_KEYS_TO_TRANSFORM), None)
+
+    def _run_cmd(self, cmd):
+        # connect
+        session = self._create_session()
+        # execute
+        response = self._winrm_run_cmd(session, cmd, env=self._env, cwd=self._cwd)
+        # create triplet from WinRM response
+        return self._translate_response(response)
+
+    def _run_ps(self, powershell):
+        # connect
+        session = self._create_session()
+        # execute
+        response = self._winrm_run_ps(session, powershell, env=self._env, cwd=self._cwd)
+        # create triplet from WinRM response
+        return self._translate_response(response)
+
+    def _multireplace(self, string, replacements):
+        """
+        Given a string and a replacement map, it returns the replaced string.
+        Source = https://gist.github.com/bgusach/a967e0587d6e01e889fd1d776c5f3729
+        Reference = https://stackoverflow.com/questions/6116978/how-to-replace-multiple-substrings-of-a-string  # noqa
+        :param str string: string to execute replacements on
+        :param dict replacements: replacement dictionary {value to find: value to replace}
+        :rtype: str
+        """
+        # Place longer ones first to keep shorter substrings from matching where
+        # the longer ones should take place
+        # For instance given the replacements {'ab': 'AB', 'abc': 'ABC'} against
+        # the string 'hey abc', it should produce 'hey ABC' and not 'hey ABc'
+        substrs = sorted(replacements, key=len, reverse=True)
+
+        # Create a big OR regex that matches any of the substrings to replace
+        regexp = re.compile('|'.join([re.escape(s) for s in substrs]))
+
+        # For each match, look up the new string in the replacements
+        return regexp.sub(lambda match: replacements[match.group(0)], string)
+
+    def _param_to_ps(self, param):
+        ps_str = ""
+        if isinstance(param, six.string_types):
+            ps_str = '"' + self._multireplace(param, PS_ESCAPE_SEQUENCES) + '"'
+        elif isinstance(param, bool):
+            ps_str = "$true" if param else "$false"
+        elif isinstance(param, list):
+            ps_str = "@("
+            ps_str += ", ".join([self._param_to_ps(p) for p in param])
```

Comment:
How do we quote the parameters here? Is this function parameter injection safe? Or we need to do additional quoting? Or it happens somewhere else?

Reply:
Quoting/escaping is done here: https://github.com/nmaludy/st2/blob/feature/winrm-runner/contrib/runners/winrm_runner/winrm_runner/winrm_base.py#L268-L269

We use the `_multireplace()` function to escape the following sequences: https://github.com/nmaludy/st2/blob/feature/winrm-runner/contrib/runners/winrm_runner/winrm_runner/winrm_base.py#L63-L80

In that dict, all keys are the "source" character, and all values are what they're replaced with.

Reply:
Thanks for clarifying.

I see `_multireplace` uses regexes. I hope it's safe :)